### ISSUE TITLE
Ensure HLSL legalization is turned on by default.

### ISSUE THREAD

**Issue by dneto0** [google/shaderc#380]
Depends on the success and maybe the form of https://github.com/KhronosGroup/glslang/pull/1070

We might just have to set a Glslang compilation flag, or run the passes ourselves.

**Comment by dneto0** [comment]
Hi @atgoo  Please take a look at this.

This is a pretty important change in the HLSL compilation flow. It would be a great learning experience as well.  I could use extra hands in maintaining it.

**Comment by dneto0** [comment]
There are several options here:
1. Compile Glslang so that it runs the optimization passes automatically.  
  * I think that requires the SPIRV-Tools and SPIRV-Headers be placed in a particular subdirectory.  But I'm not sure.   We don't want to force that because we support more flexible configurations.   So perhaps send a PR to Glslang to add the flexibility about where the sources are placed, much like Shaderc can use a SPIRV-Tools from another place. (Shaderc detects the SPIRV-Tools build target.)
2. Instead of Glslang running those passes, have Shaderc's compiler run those passes itself. Shaderc already links to SPIRV-Tools.  This is simpler from a build configuration perspective, but has a small downside that we now have the list of passes in two places.  But I expect the list to be pretty stable.   
   * Another wrinkle is that Glslang uses the dead-code-elimination (DCE) functionality from the spirv-remap code in Glslang.   I'd prefer we use the DCE functionality in SPIRV-Tools.  There is already eliminate-dead-functions, but not much of general DCE.  That's planned to come.

**Comment by antiagainst** [comment]
After https://github.com/google/shaderc/pull/387, glslang should already have been picking up SPIRV-Tools-opt.

I actually like the second option more since it gives us more control about which pass to run and having one central place for managing both legalization and optimization passes. The annoying thing is that glslang may output warnings saying input HLSL needs legalization (since glslang is not configured to include SPIRV-Tools) but we latter actually do the legalization. Maybe we need to check whether there are ways to disable such warnings (if exists). There are [APIs](https://github.com/KhronosGroup/glslang/blob/5a57ca680f5529433551ea8aea15aa6920f7b63e/glslang/MachineIndependent/localintermediate.h#L615) to query whether legalization is necessary.

**Comment by dneto0** [comment]
Shaderc already filters Glslang's messages, to make them more like other command line compilers. So maybe we filter it then.  :-)

**Comment by dneto0** [comment]
Hi @atgoo are you working on this?  Please claim it by saying so in this issue.

**Comment by atgoo** [comment]
Yes. I can't assign it to myself though.

**Comment by dneto0** [comment]
Right.  But the procedure in that case is you add a note to the issue saying you are working on it.

**Comment by dneto0** [comment]
https://github.com/google/shaderc/pull/394 did the C++ API

Now we need an update for the C API to add an option to allow a client to turn off HLSL legalization
